Sqlservertime

smootoo opened the pull request.

smootoo commented:
As requested. A new PR based on #1825 
Don't merge this one yet. I'll fix the Oracle tests first.

BTW, do we still care about the CircleCI build. It looks like it's failing on master as well. I could fix that too?

hvesalai commented:
I must say I don't know when the circleci tests are run. I have not been following them.

smootoo commented:
Just a quick update on where I am with this. I got the Oracle profile working with the existing tests, but as I was reviewing the code in the profile, I was confused as to how it could be working. Turns out it didn’t work :-) There is a gap in the test coverage. In particular around updating ResultSets and what’s happening at DST change overs. I’m increasing the test coverage and it’s highlighted some issues with the other profiles as well. I’ll work through those and update this PR. I should have something by this weekend.

smootoo commented:
Another quick update on where I am. The new tests I added highlighted a few issues, so I've been working through them. It's taken a while, but I'm getting close now. Only the DB2 and postgres tests to fix. Hopefully have them done and the code tidied up in a few days time.

hvesalai commented:
Thank you for the update. Good progress!

smootoo commented:
Almost there now. This has turned into a bit of an epic! Fixing the postgres issues were particularly tricksy.
All the builds are green now though. I'll re-review the code in its entirety and beef up some of the comments as I've been heads-down fixing one issue after another.

hvesalai commented on slick/src/main/scala/slick/jdbc/DB2Profile.scala:
```diff
@@ -180,6 +185,70 @@ trait DB2Profile extends JdbcProfile {
     }
+
+    class InstantJdbcType extends super.InstantJdbcType {
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(6) WITH TIME ZONE"
```
Why would Instant have a time zone? Instant is just time steps since `1970-01-01T00:00:00Z`, i.e. it is always Zulu time

hvesalai commented on slick/src/main/scala/slick/jdbc/DerbyProfile.scala:
```diff
@@ -240,6 +243,27 @@ trait DerbyProfile extends JdbcProfile {
+
+    class InstantJdbcType extends super.InstantJdbcType {
+      // Derby has no timestamp with timezone type and so using strings as timestamps are
+      // susceptible to DST gaps twice a year
```
this comment should not apply to Instant, which should be stored in Z-time, i.e. without DST

hvesalai commented on slick/src/main/scala/slick/jdbc/DerbyProfile.scala:
```diff
@@ -240,6 +243,27 @@ trait DerbyProfile extends JdbcProfile {
+        r.updateString(idx, v.toString)
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        s"'${value.toString}'"
```
Here the `toString` of Instant is defined as returning a string in the UTC-timezone.

hvesalai commented on slick/src/main/scala/slick/jdbc/H2Profile.scala:
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
+                      .optionalStart()
+                      .appendFraction(ChronoField.NANO_OF_SECOND, 0, 9, true)
+                      .optionalEnd()
+                      .appendOffset("+HH", "")
```
Ditto here. Instant should not be stored only in the UTC zoneoffset, which may also be implied.

hvesalai commented on slick/src/main/scala/slick/jdbc/H2Profile.scala:
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
+                      .optionalEnd()
+                      .appendOffset("+HH", "")
+                      .toFormatter()
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(9) WITH TIME ZONE"
```
As instants should always be in the UTC zone offset, storing the timezone is not necessary

hvesalai commented on slick/src/main/scala/slick/jdbc/H2Profile.scala:
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
+      override def setValue(v: Instant, p: PreparedStatement, idx: Int) : Unit = {
+        p.setString(idx, if (v == null) null else v.toString)
+      }
+      override def getValue(r: ResultSet, idx: Int) : Instant = {
```
this is different from the DerbyProfile, which uses `Instant.parse` Why are they different?

hvesalai commented on slick/src/main/scala/slick/jdbc/HsqldbProfile.scala:
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
+    }
+    class InstantJdbcType extends super.InstantJdbcType with HsqldbTimeJdbcTypeWithOffset {
+      override def sqlType = java.sql.Types.TIMESTAMP_WITH_TIMEZONE
+      override def sqlTypeName(sym: Option[FieldSymbol]) = "TIMESTAMP(9) WITH TIME ZONE"
```
same comment as with the other Instance types... no zone needed if we expect UTC as implied.

hvesalai commented on slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala:
```diff
@@ -164,12 +172,196 @@ trait JdbcTypesComponent extends RelationalTypesComponent { self: JdbcProfile =>
+        java.sql.Types.TIMESTAMP
+      }
+      override def setValue(v: Instant, p: PreparedStatement, idx: Int) : Unit = {
+        p.setTimestamp(idx, Timestamp.from(v))
```
This looks nice. Is there a reason why the other Profiles don't use this method?

hvesalai commented on slick/src/main/scala/slick/jdbc/DB2Profile.scala:
```diff
@@ -176,6 +181,30 @@ trait DB2Profile extends JdbcProfile {
+        r.updateString(idx, v.toString)
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        s"'$value'"
```
please add `.toString`

hvesalai commented on slick/src/main/scala/slick/jdbc/JdbcTypesComponent.scala:
```diff
@@ -164,12 +172,196 @@ trait JdbcTypesComponent extends RelationalTypesComponent { self: JdbcProfile =>
+        p.setTimestamp(idx, Timestamp.from(v))
+      }
+      override def getValue(r: ResultSet, idx: Int) : Instant = {
+        r.getTimestamp(idx) match {
```
Also looks very clean.

hvesalai commented on slick/src/main/scala/slick/jdbc/PostgresProfile.scala:
```diff
@@ -255,12 +258,191 @@ trait PostgresProfile extends JdbcProfile {
+      implicit val max : Instant = Instant.MAX
+      implicit val serializeFiniteTime : (Instant => String) =  _.toString
+      implicit val parseFiniteTime : (String => Instant) = {
+        LocalDateTime.parse(_, formatter).toInstant(ZoneOffset.UTC)
```
Is this really safe? Here you convert the string from the database to the Local timezone and then to Instant. As I understand it, you would get wrong time if the system timezone is not UTC and the time didn't contain a timezone (which would be OK for instant since it is implicitly always in UTC)

smootoo commented:
I wrote this as you were making more comments. In summary though, I think the timestamp is mostly the wrong datatype for Instants as DST shifts may be applied to it depending on configuration.

The timestamp (without time zone) database type is handled different in different db backends (and jdbc implementations). Postgres for example stores timestamps as UTC and so they are the datatype for Instants.
Depending on timezone configurations in the database backend and on the client JVM, DST may be applied to the timestamp. In a timezone where DST is in effect, then an hour may be gained or lost during the changeover. For example, in ‘London/Europe’, if you try to write the instant ‘2018-03-25T01:37:40Z’ into a timestamp column, it will come back from the db as ‘2018-03-25T02:37:40Z’.
This is acceptable and probably intended for LocalDateTime, but Instants need to be continuous and always increasing, with a larger value being later on the timeline.
Where timestamps with time zone are supported by the db backend, then writing Instants as timestamps with timezone ‘UTC’ or offset ‘+00:00’ gives us the characteristics we need.

I’ll write up something like the above in a big comment so that people know what’s going on and why.

Reviewing this PR is the time to pick up on this though, so thanks for looking. We now have a comprehensive set of tests try alternative implementations to make sure there are no edge cases where values aren’t round-tripped correctly.

smootoo commented:
I agree with you that for Instants, `2018-03-25T01:37:40Z` must stay unchanged whatever the locally configured time zone is.
But if that same Instant is converted into a `LocalDateTime` (i.e. a type without a timezone) and that is stored in a database that is configured with a local time zone of `London/Europe` into a database column of type `TIMESTAMP` then because this time is in the local DST gap (the clocks have gone forward an hour), the db will turn that time into `2018-03-25T02:37:40`. This is what I meant by probably intended behaviour for `LocalDateTime`, that any locally enforced DST gaps are applied, but the value still doesn't have a timezone.

hvesalai commented:
Ah ok, so then we need just big block letters saying *Caveat emptor DST*

smootoo commented:
Indeed. There are dragons here everywhere! Like I said, I'll stick in a big comment explaining what's going on. I personally couldn't think of a use for LocalDateTime that isn't better served with another time type. But I guess it supports some legacy behaviours.

hvesalai commented:
Instead of a big comment, should we invest in a new page on the documentation?

smootoo commented:
Sure. I'll draft something at the weekend. OK?

hvesalai commented:
Yes, that sounds good!

hvesalai commented:
Also, if you can say when the code is ready for a new round of code-review, I can then review and add comments.

smootoo commented:
Sorry I updated this later that I had planned. Real life got in the way over the last week!

I added a page of documentation. The code is ready for review now.

The PR has been open in various forms for a long time and has had many different contributors. A fresh pair for eyes would good now. Appreciate any comments you may have.

hvesalai commented:
@smootoo can you rebase your PR, there is some conflicts

hvesalai commented on slick/src/main/scala/slick/jdbc/H2Profile.scala:
```diff
@@ -123,6 +127,32 @@ trait H2Profile extends JdbcProfile {
       override def hasLiteralForm = true
     }
+    override val instantType : InstantJdbcType = new InstantJdbcType {
+      val formatter = new DateTimeFormatterBuilder()
```
Can you add a comment why we need the formatter here. DB2 and Derby use `Instant.parse`, which can't this?

smootoo replied:
Comment added

hvesalai commented on slick/src/main/scala/slick/jdbc/HsqldbProfile.scala:
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
+
+      override def getValue(r: ResultSet, idx: Int): LocalTime = {
+        // Fixes an issue caused because Hsqldb outputs the time in a non-standard format
+        // not adding trailing zeros to the hours. For example: '2:14:41.421' instead of '02:14:41.421'
```
comment speaks about "trailing" (after something), maybe it means "leading" (before something)

smootoo replied:
Comment fixed. You're right. Is should have been leading.

hvesalai commented on slick/src/main/scala/slick/jdbc/HsqldbProfile.scala:
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
+        // not adding trailing zeros to the hours. For example: '2:14:41.421' instead of '02:14:41.421'
+        r.getString(idx) match {
+          case null => null
+          case isoTime if isoTime.indexOf(':') == 2 => LocalTime.parse(isoTime)
```
Why is the formatter on line 162 not used?

smootoo replied:
The formatter on line 162 is for offset times. It has a non-optional offset component and so can't be used for LocalTime. The default LocalTime parser (which is used here) works fine though.

hvesalai commented on slick/src/main/scala/slick/jdbc/HsqldbProfile.scala:
```diff
@@ -125,8 +128,160 @@ trait HsqldbProfile extends JdbcProfile {
+      /**
+        * Add offset padding e.g. +1:00 becomes +01:00.
+        */
+      def offsetConvertHsqldbToISO(hsqldbString : String) : String = hsqldbString.takeRight(5).toArray match {
```
Why are these needed? They look a bit suspicious to me. Couldn't you just use a `DateTimeFormatter` instead? 

Please add a comment on why a `DateTimeFormatter` cannot be used.

smootoo replied:
The comment on lines 157 to 159 explain why these are needed in addition to a `DateTimeFormatter` This whole trait is for handling that the offsets don't have leading zeros.

hvesalai replied:
Ok

hvesalai commented on slick/src/main/scala/slick/jdbc/OracleProfile.scala:
```diff
@@ -324,13 +335,154 @@ trait OracleProfile extends JdbcProfile {
+    // So, store them at UTC timestamps, otherwise the JDBC layer might attempt to map them
+    // and with DST changes, there are some times which will be unrepresentable during the switchover
+    class LocalDateTimeJdbcType extends super.LocalDateTimeJdbcType {
+//      private[this] val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS x")
```
remove the commented out code if there is no reason to keep it. If there is a reason, document it.

smootoo replied:
Commented out code removed

hvesalai commented on slick/src/main/scala/slick/jdbc/OracleProfile.scala:
```diff
@@ -404,3 +556,180 @@ object OracleProfile extends OracleProfile {
 }
+
+
+/* The following code will be used when the native implementation of
```
what does this comment mean? "native implementation of ... are natively implemented"

smootoo replied:
That comment was left over from one of the intermediate commits. Removed.

hvesalai commented on slick/src/main/scala/slick/jdbc/SQLServerProfile.scala:
```diff
@@ -248,29 +255,143 @@ trait SQLServerProfile extends JdbcProfile {
+      }
+      override def valueToSQLLiteral(value: Instant) = {
+        value match {
+          case null =>
```
this is the only profile that handles `value == null`. Why don't the other do it?

smootoo replied:
The `case null =>` couldn't be reached. They've been removed.

hvesalai commented:
Ok. Now there is a plethora of comments. Can you go through them all and reply / make changes as needed.

Also it would be very good if @tminglei had time to read the Postgres code here.

smootoo commented:
I can't reply directly to your comment on the DB2Profile.
I don't understand that comment. Could you elaborate?

smootoo commented on doc/src/datetimetypes.md:
```diff
@@ -0,0 +1,25 @@
+That means that you get back the same value that was stored, irrespective of the configuration of the
+database or the JVM.
+
+2. It should be clear to you as a user of the database what those values are. This may be with an appropriate database type, if one is available. If not, a varchar with a "stringified" value will be used.
```
Removed my made up word.
Elaborated on the comparability of the string values.

smootoo commented on doc/src/datetimetypes.md:
```diff
@@ -0,0 +1,25 @@
+
+The different java.time types have important characteristics. E.g `Instant`s need to be continuous and always increasing,
+with a larger value being later on the timeline.
+The timestamp (without time zone) database type is handled different in different db backends (and jdbc implementations). Postgres, for example, stores timestamps as UTC and so they are approproriate to be the datatype for Instants. Depending on timezone configurations in the database backend and on the client JVM, DST may be applied
```
I did mean for db others than postgres. Updated.
Added explanation for DST and why it's a problem.

smootoo commented:
@hvesalai I've updated based on your comments. Thanks.
I don't think this needs a rebase. I can merge as is, cleanly.

hvesalai commented:
Added two more requests to review coments.

smootoo commented on slick-testkit/src/codegen/scala/slick/test/codegen/GenerateRoundtripSources.scala:
```diff
@@ -79,7 +79,7 @@ class Tables(val profile: JdbcProfile){
     def pk = column[Int]("pk")
     def pk2 = column[Int]("pk2")
-    def pkpk = primaryKey( "", (pk,pk2) ) // pk column collision
+    def pkpk = primaryKey( "primary_key2", (pk,pk2) ) // pk column collision
```
PR #1891 has been raised

smootoo commented on doc/src/datetimetypes.md:
```diff
@@ -0,0 +1,25 @@
+One java.time type to be careful of is `LocalDateTime`, that behaves most like the legacy `java.sql.Timestamp` and is mapped into a Timestamp database type. For these, the exact behaviour for DST related can depend on the environment configuration of the database or the JVM. And so is the one type that can break the "roundtripping" rule if the value being stored falls into a DST gap for a given e ...
+
+For example, if your environment is configured to be `Europe/London`, and you try to write the `LocalDateTime` `2018-03-25T01:37:40` into a database that performs DST shifts, such as Oracle, it will come back from the db as `2018-03-25T02:37:40`. This may be the desired behaviour, and is what happens with some databases for a Timestamp without timezone type. If this isn't the desired behaviour, t ...
+
```
Documentation updated

hvesalai commented:
Almost there. Now if you can squash all the commits to one and rebase, then I can hit the merge button. Now this PR still contains the same changes as in #1891, but after a rebase they should be gone.

hvesalai commented:
Here goes the planet! Thank you!

nafg commented:
Well... this broke my app, and not at compile time...

I have different columns of different SQL types in various places that are mapped to Instant. I don't see a workaround other than doing one of the following (which I probably should do anyway, but that doesn't mean I would prefer less overhead right now):

1. Convert the old column types
2. Use something like a tagged type, so that the column type includes the underlying type somehow but values can still be used as the mapped type (essentially along the lines of `Instant with Phantom[Long]`)

hvesalai commented:
@nafg there was somebody with identical problem I think. Just can't remember where that discussion is.
